### PR TITLE
transition to S3 methods

### PR DESCRIPTION
**Pull request by simonpcouch**
Trying to keep these PRs as small as possible so the review isn’t too cumbersome—this PR converts the modeling framework to S3 but generally holds off on adjusting inputs/outputs or carry out much refactoring! The output of the modeling function (`fe_bayes_lm`) gets a class, `fe_bayes_lm`, and then we can implement `predict` and `autoplot` (and eventually `print`) wrt to that class.

I moved the plot methods to their own file, so the diff is a bit harder to track, but the gist is that the plotting function now makes use of ggplot2’s `autoplot` generic function. The user can use the `type` argument to determine which of the three plots they’d like to return.

Feel free to adjust the names of the plot types (`"intercepts"`, `"slopes"`, `"competition"`) as you see fit. I’m also all ears on adjusting the name of the modeling function/class!

Also, I have some local changes (resulting from global search + replace) to the paper `.Rmd`—I’m glad to push those if appreciated, though anticipate yall probably want to hold off on paper edits until the API has solidified a bit more.

Going to wait to implement a `print.fe_bayes_lm` method until the structure of that object is a bit more fleshed out in a future PR.

### CODE REVIEW

File: R/model_functions.R [hunk 1]
```diff
@@ -93,7 +93,11 @@ fit_bayesian_model <- function(focal_vs_comp, prior_param = NULL, run_shuffle =
     V_star = V_star,
     sp_list = sp_list
   )
-  return(posterior_hyperparameters)
+
+  structure(
+    posterior_hyperparameters,
+    class = c("fe_bayes_lm", class(posterior_hyperparameters))
+  )
```
Comment (rudeboybert) [inline R/model_functions.R:100]: huh, so this is all it takes to define an S3 class. Interesting 🤔

File: R/model_functions.R [hunk 2]
```diff
@@ -125,17 +133,18 @@ fit_bayesian_model <- function(focal_vs_comp, prior_param = NULL, run_shuffle =
 #' # and growth data to compare to
 #' data(posterior_param_ex, ex_growth_df)
 #'
-#' predictions <- focal_vs_comp_ex %>%
-#'   predict_bayesian_model(posterior_param = posterior_param_ex) %>%
+#' predictions <- posterior_param_ex %>%
+#'   predict(focal_vs_comp = focal_vs_comp_ex) %>%
```
Comment (rudeboybert) [inline R/model_functions.R:137]: now that I think about it `predict.lm()` has the model as the first argument and the `newdata` as the second as well. I did wonder about this, given the tidyverse's propensity to have the data frame argument always come first

File: R/plot.R [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+#' # Plot beta_0, growth intercepts
+#' autoplot(posterior_param_ex, type = "intercepts")
+#'
+#' # Plot beta_dbh, growth-dbh slope
+#' autoplot(posterior_param_ex, type = "slopes")
+#'
+#' # Plot lambdas, competition coefficients
+#' autoplot(posterior_param_ex, type = "competition")
+#' @export
+autoplot.fe_bayes_lm <- function(object,
```
Comment (rudeboybert) [inline R/plot.R]: ah, so `autoplot.SOMETHING()` is the ggplot2 equivalent of the `plot.SOMETHING()` generic in base R. This approach is very clean!

**Review by rudeboybert — APPROVED** [review]
Thanks for all your comments and outlining the decision points.

- Yep, I like having bite-sized PR's too. 
- I am def down with not refactoring our functions right now (if ever before submission, done is better than perfect after all ;)).
- As for the `.Rmd`, if you have the local changes down, could you start a branch off this branch sooner rather than later? Just so we see what's coming up in terms of writing?
- I renamed the class from `fe_bayes_lm()` to `comp_bayes_lm()` to be more specific to our "competition model"

Please merge at your leisure!

File: R/model_functions.R (hunk 1 above)
Comment (simonpcouch) [reply R/model_functions.R:100]: Hahaha yup! Some folks even just `class(x) <- "newclass"`. :-)

**Comment by simonpcouch** [comment]
Thanks for the fixes + review!

Will make a new branch and push those changes to the paper—again, only resulting from global find + replace, and will likely require some workshopping to get working probably.

I'll be shooting for aligning input/output of these methods with tidy data principles by our meeting on Monday.

**Pull request merged**